File: articles/calculus-for-systems-modeling/articles/the-fundamental-theorem-of-calculus/fortran/fundamental_theorem_audit.f90
program fundamental_theorem_audit
  implicit none
  real(8), dimension(9) :: times
  real(8) :: accumulated_rate, endpoint_difference, residual, dt
  integer :: i
  times = (/0.0d0,0.25d0,0.5d0,0.75d0,1.0d0,1.25d0,1.5d0,1.75d0,2.0d0/)
  accumulated_rate = 0.0d0

  do i=1,8
    dt = times(i+1) - times(i)
    accumulated_rate = accumulated_rate + 0.5d0 * (rate_value(times(i)) + rate_value(times(i+1))) * dt
  end do

  endpoint_difference = state_value(times(9)) - state_value(times(1))
  residual = endpoint_difference - accumulated_rate

  print '(A)', 'interval_start interval_end endpoint_difference accumulated_rate residual'
  print '(F8.3,1X,F8.3,1X,F14.6,1X,F14.6,1X,F14.6)', times(1), times(9), endpoint_difference, accumulated_rate, residual

contains
  real(8) function state_value(t)
    real(8), intent(in) :: t
    state_value = 50.0d0 + 2.0d0*t + 3.0d0*sin(t)
  end function

  real(8) function rate_value(t)
    real(8), intent(in) :: t
    rate_value = 2.0d0 + 3.0d0*cos(t)
  end function
end program

File: articles/calculus-for-systems-modeling/articles/the-fundamental-theorem-of-calculus/cpp/fundamental_theorem_audit.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

double state_value(double t){ return 50.0 + 2.0 * t + 3.0 * std::sin(t); }
double rate_value(double t){ return 2.0 + 3.0 * std::cos(t); }

int main(){
  std::vector<double> times{0,0.25,0.5,0.75,1,1.25,1.5,1.75,2};
  double accumulated_rate = 0.0;

  for(size_t i=0;i+1<times.size();++i){
    double dt = times[i+1] - times[i];
    accumulated_rate += 0.5 * (rate_value(times[i]) + rate_value(times[i+1])) * dt;
  }

  double endpoint_difference = state_value(times.back()) - state_value(times.front());
  double residual = endpoint_difference - accumulated_rate;

  std::cout<<std::fixed<<std::setprecision(12);
  std::cout<<"interval_start,interval_end,endpoint_difference,accumulated_rate,residual\n";
  std::cout<<times.front()<<","<<times.back()<<","<<endpoint_difference<<","<<accumulated_rate<<","<<residual<<"\n";
}
